### ISSUE TITLE
Have you noticed a big unbalance between gpus with InplaceSyncBN?

### ISSUE THREAD

**Issue by CoinCheung** [mapillary/inplace_abn#83]
With pytorch nn.BatchNorm2d and nn.DataParallel, the memory difference between two gpus is less than 1G. However, with InplaceSyncBN and nn.DataParallel， the memory gap is enlarged to almost 2G. What is the cause of this and how could I avoid it?

**Comment by rotabulo** [comment]
@CoinCheung the new layer is not supposed to be used with `nn.DataParallel`, but `nn.DistributedDataParallel`. Could you try with the latter?

**Comment by CoinCheung** [comment]
Hi, must I use `nn.DistributedDataParallel` even if I am working on a single machine with 4 gpus?
If I use the naive `nn.DataParallel`, apart from the memory unbalance, will I suffer other dysfunctions (for example, with `nn.DataParallel` the bn is actually not synchronized at all)?

**Comment by rotabulo** [comment]
@CoinCheung you can use `nn.DistributedDataParallel` also within the same node, and it's actually faster than `nn.DataParallel` (see, https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L55).

**Comment by CoinCheung** [comment]
I noticed that there is a deeplab.py in the modules/ foler. Have you ever trained it with sync-abn? It seems that pytorch does not support training based on pixel wise cross entropy.

**Comment by rotabulo** [comment]
@CoinCheung there you find our implementation of the deeplab head, which we used to train our segmentation model with synchronized iABN. Pytorch supports pixel-wise cross-entropy via `torch.nn.CrossEntropyLoss`.

**Comment by CoinCheung** [comment]
Would you please show me how to train it, I cannot make distributed training work, a simplified version of my code is like this: 
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import time

import argparse


def parse_args():
    parse = argparse.ArgumentParser()
    parse.add_argument(
            '--local_rank',
            dest = 'local_rank',
            type = int,
            default = -1,
            )
    return parse.parse_args()


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3,
            64,
            kernel_size = 3,
            stride = 2,
            padding = 1,)
        self.conv2 = nn.Conv2d(64,
            256,
            kernel_size = 3,
            stride = 2,
            padding = 1,)
        self.conv3 = nn.Conv2d(256,
            19,
            kernel_size = 3,
            stride = 2,
            padding = 1,)
        self.linear = nn.Linear(512, 10)

    def forward(self, x):
        H, W = x.size()[2:]
        x = self.conv1(x)
        x = self.conv2(x)
        logits = self.conv3(x)
        logits = F.interpolate(logits, (H, W), mode='bilinear')
        return logits

def train():
    args = parse_args()

    torch.cuda.set_device(args.local_rank)
    torch.distributed.init_process_group(
                backend='nccl',
                init_method='env://',
                )
    net = Net()
    net.train()
    net.cuda()
    #  net = nn.DataParallel(net)
    net = nn.parallel.DistributedDataParallel(net,
            device_ids = [args.local_rank, ],
            output_device = args.local_rank
            )

    optim = torch.optim.SGD(
            net.parameters(),
            lr = 1e-3,
            momentum = 0.9,
            weight_decay = 5e-4)
    criteria = nn.CrossEntropyLoss()

    for i in range(10000):
        img = torch.randn(2, 3, 768, 768).cuda()
        lb = torch.randint(0, 18, [2,768,768]).cuda()
        optim.zero_grad()
        out = net(img)
        #  out = out.permute(0,2,3,1).contiguous().view(-1, 19)
        #  lb = lb.view(-1, )
        #  print(out.size())
        #  print(lb.size())
        loss = criteria(out, lb)
        loss.backward()
        optim.step()

if __name__ == "__main__":
    train()
```

**Comment by CoinCheung** [comment]
Hi, 

I finally tried to make my code work. Here is what I am still not sure: If I use inplace-ABN but work on only one gpu, which the model be trained the same way as with nn.BatchNorm2d, or should I modify my model and use nn.BatchNorm2d in the one gpu scenario?

**Comment by rotabulo** [comment]
@CoinCheung inplace-ABN per se has no relation with number of gpus, it's a layer like any other. If you meant the synchronized version of inplace-ABN, then it works also with a single GPU.

**Comment by CoinCheung** [comment]
Hi, 

Is there anything I should take care of when I use sync-inplace-abn apart from replacing `nn.BatchNorm2d` and `nn.ReLu(inplace=True) with `inplace_abn.modules.bn.InPlaceABNSync`? I just setup my  distributed program like this: 
```python
    args = parse_args() # this contains local_rank
    torch.cuda.set_device(args.local_rank)
    dist.init_process_group(
                backend='nccl',
                init_method='tcp://127.0.0.1:33488',
                world_size=2,
                rank=args.local_rank
                )
```
Then  I wrap my model with ordinary method:
```python
    net = nn.parallel.DistributedDataParallel(net,
            device_ids = [args.local_rank, ],
            output_device = args.local_rank
            )
``` 
My dataset is manipulated with distributed sampler: 
```python
    sampler = torch.utils.data.distributed.DistributedSampler(ds)
    dl = DataLoader(ds,
                    batch_size = batchsize,
                    shuffle = False,
                    sampler = sampler,
                    num_workers = n_workers,
                    pin_memory = True,
                    drop_last = True)
``` 
In the command line I train like this:
```
$ python -m torch.distributed.launch --nproc_per_node=2 train.py
```
Will these settings be enough to make the sync version of inplace-abn work, or shall I include other operations?

**Comment by rotabulo** [comment]
@CoinCheung I am closing the issue, since the unbalanced training was due to using `DataParallel` instead of `DistributedDataParallel`
